optax/assignment/_hungarian_algorithm_test.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2024 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+'''Tests for the Hungarian algorithm.'''
+
+from absl.testing import absltest
+from absl.testing import parameterized
+from jax import random, numpy as jnp
+import scipy
+import optax
+
+
+class HungarianAlgorithmTest(parameterized.TestCase):
+
+  @parameterized.product(
+    n=[0, 1, 2, 4, 8, 16],
+    m=[0, 1, 2, 4, 8, 16],
+  )
+  def test_hungarian_algorithm(self, n, m):
```

Comment:
It would be great to see whether this algorithm can support usual transformations (like jitting or vmap).

Side note beyond this PR:
It could be great to have this building block differentiable. To be exact: there is no way the current outputs can be differentiable (they are ints). But the resulting cost could be differentiable. The advantage is that people may then parameterize the cost matrix, run the algorithm to get the optimal cost and then differentiate through the algorithm to further refine their cost matrix.
This would not be super hard actually to let your function return the optimal cost so that a user could easily make it differentiable. 
Anyway, it's not very important, so you can dismiss this comment

Reply:
Currently, the algorithm has 2 occurrences of [lax.while_loop](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.while_loop.html), which is not reverse-mode differentiable.

In theory, these could be replaced with a [bounded while_loop](https://github.com/jax-ml/jax/issues/2469). I don't think that's been ported [from jaxopt](https://github.com/google/jaxopt/blob/main/jaxopt/_src/loop.py) yet, has it?

On the other hand, it seems to me that, for [almost all](https://en.wikipedia.org/wiki/Almost_all) inputs and infinitesimal perturbations, the (integer) solution indices remain unchanged, and thus differentiating the solution cost is equivalent to simply differentiating `cost_matrix[i_solution, j_solution]`.

Reply:
> has 2 occurrences of lax.while_loop

Ok, I should have been more careful, this is indeed an impediment for an efficient differentiation.
Anyway, we may think about that later, the current implementation should already be useful for many use cases.
And thanks for checking the interoperability with jit and vmap!